File: problems/p209/p209.go
package p209

func minSubArrayLen(target int, nums []int) int {
	var l = len(nums)
	var result = l + 1
	var val = 0
	var left = 0
	for i := 0; i < l; {
		if val >= target {
			val -= nums[left]
			left++
			result = min(result, i-left+1)
			continue
		}
		val += nums[i]
		i++
	}
	for val >= target {
		val -= nums[left]
		left++
		result = min(result, l-left+1)
	}
	if result > l {
		return 0
	}
	return result
}

func min(i, j int) int {
	if i < j {
		return i
	}
	return j
}

File: problems/p209/p209.py
from typing import List
import unittest


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        num_len = len(nums)
        result = num_len+1
        left = 0
        win_sum = 0
        right = 0
        while right < num_len:
            if win_sum >= target:
                win_sum -= nums[left]
                left += 1
                result = min(result, right-left+1)
                continue
            win_sum += nums[right]
            right += 1
        while win_sum >= target:
            win_sum -= nums[left]
            left += 1
            result = min(result, num_len-left+1)
        if result > num_len:
            return 0
        return result


class Test(unittest.TestCase):
    def setUp(self) -> None:
        self.s = Solution()

    def test_example_1(self):
        self.assertEqual(self.s.minSubArrayLen(7, [2, 3, 1, 2, 4, 3]), 2)

    def test_example_2(self):
        self.assertEqual(self.s.minSubArrayLen(4, [1, 4, 4]), 1)

    def test_example_3(self):
        self.assertEqual(self.s.minSubArrayLen(
            11, [1, 1, 1, 1, 1, 1, 1, 1]), 0)


if __name__ == "__main__":
    unittest.main("p209")
